main/src/ca/uwaterloo/flix/api/lsp/FormattingOptions.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2025 gwydd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp
+
+import org.json4s.MonadicJValue.jvalueToMonadic
+import org.json4s.JObject
+import org.json4s.JsonAST.{JBool, JInt, JValue}
+import org.json4s.JsonDSL._
+
+import org.eclipse.lsp4j
+
+case class FormattingOptions(
+                              tabSize: Int,
+                              insertSpaces: Boolean,
+                              trimTrailingWhitespace: Option[Boolean] = None,
+                              insertFinalNewline: Option[Boolean] = None,
+                              trimFinalNewlines: Option[Boolean] = None
+                            ) {
+
+  def toJSON: JObject = {
+    var obj: JObject = ("tabSize" -> tabSize) ~ ("insertSpaces" -> insertSpaces)
+
+    // FormattingOptions that are optional and were introduced in version 3.15 (unused)
+    trimTrailingWhitespace.foreach(v => obj = obj ~ ("trimTrailingWhitespace" -> v))
+    insertFinalNewline.foreach(v => obj = obj ~ ("insertFinalNewline" -> v))
+    trimFinalNewlines.foreach(v => obj = obj ~ ("trimFinalNewlines" -> v))
```

Comment:
The above code is wrong? I think it is just no-op. Because this is an immutable structure.